fix: handling of language code copied #674

mikbry opened the pull request.

kustomzone commented:
Not well versed in React besides installing and testing countless apps.
I see you're stepping thru dom nodes: `const start = node.position?.start.offset || 0;`

Haven't tested besides a few queries to new model (Starling-LM-7B)
A clean reinstall, i'm getting correct output on the few Python code blocks, but JS is only working in 2 out of 5 cases.

[JS]  'use strict;'  . . . [object Object],; 

Not sure where to start debugging...
Because i had asked about React with Tauri, then asked about React's clipboardData mime type i got this suggestion.

"The `handleCopyCode` function uses the `clipboard.writeText()` API from Tauri's clipboard module with the
 'application/x-python' mime type to write the code to the clipboard."
 
This [LLM at least] seems to assume it's compiled.
 
py:  'text/x-python',
pyc:  'application/x-python-code'

 Weirdly, the first mime type list i got from random google repo has:  js:  'application/x-javascript'
But the spec says: text/javascript  (Specifications: HTML and RFC 9239)

It's unlikely that Tauri's clipboard module is outputting JS in different mime types, so most likely the node depth is changing in some cases.

mikbry commented:
Could copy paste here messages with faulty js. I think the markdown is not well formatted. It is not a DOM node, but a Hast node, the parsed node.

kustomzone commented:
'use strict;' . . . [object Object],;
 
 `<div class="w-full font-sans text-sm"> 
<div><strong class="font-semibold text-neutral-950 dark:text-neutral-100">Use strict mode</strong>: Enable JavaScript's Strict Mode by adding the following line at the beginning of your script:</div>
<div class="relative"><pre class="font-code flex items-start overflow-x-auto rounded-sm border-neutral-200  bg-neutral-100 text-sm dark:border-neutral-800 dark:bg-neutral-900 [&amp;>code.hljs]:bg-transparent [&amp;>code.hljs]:p-0"><div class="m-0 flex w-full flex-col p-0"><div class="flex w-full flex-row items-center justify-end gap-1 bg-card"><p class="flex-grow pl-4 text-xs text-card-foreground">javascript</p><button class="inline-flex items-center justify-center whitespace-nowrap font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-8 rounded-md px-3 text-xs" aria-label="Copy code to clipboard" title="Copy code to clipboard"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-clipboard h-4 w-4 text-muted-foreground"><rect width="8" height="4" x="8" y="2" rx="1" ry="1"></rect><path d="M16 4h2a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h2"></path></svg></button></div><code class="hljs language-javascript  w-full flex-shrink flex-grow"><span class="hljs-meta">'use strict'</span>;
</code></div></pre></div>
<div>This enforces stricter parsing and error handling, which can help catch common coding mistakes early on.</div>
 </div>`

mikbry commented:
I made a fix. It seems to be an issue with Regex, it works using https://regex101.com/ and javascript selected, but not in the app. the last '\n' is guilty...
Instead of html dump, you could send me the full content of the message using copy to clip board or edit and copy the input the toolbar at the bottom of the message.
<img width="159" alt="Screenshot 2024-04-03 at 11 05 29" src="https://github.com/Opla/opla/assets/395718/33ab8ba8-8131-471f-95b3-07fddacacb0d">
<img width="750" alt="Screenshot 2024-04-03 at 11 05 13" src="https://github.com/Opla/opla/assets/395718/54f47f3b-85f3-4d81-ae31-d3626ce2729a">
Using this content I could edit a test message and past your faulty content in my dev env.
I tested with your first not-working example.

kustomzone commented:
Thanks. I assumed the message clipboard would also be funky. (it worked ok)
Unfortunately the textarea isn't extendable and only 2 lines high, so the scrollbar is not usable. Could it be set as: `div { resize: vertical; overflow: auto; }` ..so we can select partial text, as LLMs like Starling are very verbose atm.
Atm, the main message clipboard button shows tick, but doesn't reset like the code boxes, so we can't select it again if needed.
Also "reuse" icon is available on LLM output text but not on our own prompts.

The regex fix works fine, except it's indenting when the codeblock is in a numbered instruction list (therefore indented in the message). It wasn't doing this with `.innerText`.

Last example is not in a numbered list so is correct:  [indent.txt](https://github.com/Opla/opla/files/14851984/indent.txt)

mikbry commented:
I fixed the indenting problem, I think.
I made some changes to resize the textarea but it is not working correctly in the conversationlist, need to create another issue.
"Reuse" icon yep you are right, but I replicated ChatGPT toolbar ;-), should be in another issue
Regarding inconstancy of copy to clipboard button behaviour, I added a separate component in PR #690. I will fix that for MessageView in another PR.

kustomzone commented:
Haven't tried ChatGPT as they took the "play money" off me before i got a chance to test.
Comparing with [Jan](https://github.com/janhq/jan), they have the same buttons, but reuse only for the last output.
I don't see the indent fix so i'll check back for it.

mikbry commented:
I checked Jan and it behaves as ChatGPT and Opla, edit your prompt and select "save & submit" to reuse, no "reuse" button.
I created 2 issues: #696 and #695

mikbry commented:
"For prompt reuse, would it need a temperature setting / slider to adjust the seed?"
There are settings on the left panel, choose Parameters. not fully tested, but it should be applied to next completion call.

Did my fix for indenting works ?

kustomzone commented:
Still indenting with the 'merge conflicts' commit, except now it has `<|end_of_turn|>` at the end of each output. 
It looked like you pushed the textarea resize test to the 'non indented code' commit.

mikbry commented:
Thanks @kustomzone for your tests & patience ;-)
I reverted some changes, and merged this PR as the main issue is fixed. As It started to be a mess. A PR shouldn't be open for more than 2 day.
I created an issue for indentation bug #707, and I fix it ASAP
I don't have  `<|end_of_turn|>` at the end of each output. Could you confirm it is still there in the main branch. If yes I will create an issue and fix it.
